### ISSUE TITLE
Remote Filesystem Support (Request)

### ISSUE THREAD

**Issue by miversen33** [nvim-telescope/telescope-file-browser.nvim#89]
**Is your feature request related to a problem? Please describe.**
As of current, there are quite a few different local file system browser plugins for Neovim. There is really only 1 remote filesystem browser and that is the built in NetRW. NetRW has its flaws, namely being its counterintuitive approach to navigating files (It is core vim).

Telescope File Browser is an amazing telescope plugin, and I feel it could move even further ahead of other filesystem plugins by integrating ssh support.

**Describe the solution you'd like**
In a perfect world, we would be able to open a remote filesystem via neovim directly (which would use netrw) and then browse/manage said filesystem via the telescope-file-browser plugin

**Describe alternatives you've considered**
[NetRW](http://www.drchip.org/astronaut/vim/index.html#NETRW), as described above, is the only one that I am familiar with that supports remote file systems. It is... not amazing, though it is currently the only tool for the job.
[Distant.nvim](https://github.com/chipsenkbeil/distant.nvim) is another alternative, though I haven't played with it much as it depends on `distant` being installed on the remote client and that is not always an option (I am adverse to installing IDE dependencies on production machines)

**Comment by fdschmidt93** [comment]
As cool as that feature would be, it sounds largely out of scope of the project, or otherwise put, sad to say I'm not the right person for the job.

I've never used netrw (isn't it unbearably slow?) for that matter and on the few ssh connections I maintain every once in a while synchronize my dotfiles.

What would be a way to support this feature it the lightest way possible?

> In a perfect world, we would be able to open a remote filesystem via neovim directly (which would use netrw) and then browse/manage said filesystem via the telescope-file-browser plugin

That would be nice indeed.

**Comment by miversen33** [comment]
I'm not terribly familiar with lua (or neovim development) but I was able to get a (very) jank finder in place as a sort of POC to show it is _possible_. Below is the (awful) lua code I used. I am going to poke around a bit with this this weekend and see if I can 
a) Make it actually usable (it appears I am going to have to poke around with NetRW global variables and path reading)
b) Clean it up so it doesn't look awful.

There are also some design questions that should be asked. Namely
- How do you/we/the user want to designate they are searching a remote filesystem? The reason the `nvim-telescope` plugin doesn't currently work for remote filesystems is because it uses the [Plenary Path](https://github.com/nvim-telescope/telescope-file-browser.nvim/blob/e65a5674986dcaf27c0fd852b73f5fc66fa78181/lua/telescope/_extensions/file_browser/finders.lua#L16) tool which is designed explicitly for local file systems. On that note, if a user opens the nvim-telescope extension, should the decision to examine a remote filesystem be based on the "cwd" (which vim doesn't track if you are remote) or the current file they are in? Or should it be its own finder altogether (my recommendation).

Some jank lua code as a POC
```lua
local pickers = require("telescope.pickers")
local finders = require("telescope.finders")
local conf····= require("telescope.config").values
local sys·····= require("plenary.job")

--- TESTED AND WORKS

local user = ""
local host = ""
local remote_args = { "" .. user .. "@" .. host, "ls","/tmp" }

local get_remote_files = function()
    local job = sys:new({
        command = remote_command,
        args = remote_args,
    })
    job:start()
    job:join()
    return job:result()
end

local create_remote_finder = function(opts)
    opts = opts or {}
    pickers.new(opts, {
        prompt_tile = "Remote Files",
        finder = finders.new_table({
            results = get_remote_files()
        }),
        sorter = conf.generic_sorter(opts)
    }):find()
end

create_remote_finder()
```

My current recommendation is (if this is implemented) that you check the current open file to see if it starts with `sftp://` or `scp://` and if it does, you select a new finder with the remote search ability.

**Comment by fdschmidt93** [comment]
Sorry for the late reply.

In brief: As you've mentioned, `plenary.nvim` in all likelihood just fails when working via remote and with it, pretty much anything useful from `telescope-file-browser`. I'm also not necessarily convinced that running commands etc remotely is the way to go as that's always going to be overhead on the user (install required binaries) and doesn't necessarily integrate well with other plugins.

One things that comes to mind, that (at least for Unix users), could be very easily and seamlessly wrapped is `rsync`.

In other words, https://github.com/KenN7/vim-arsync with the full lua treatment once lua autocommands land could be very ergonomous. Then you could rsync remote to local `/tmp` or project folder, respectively, have a session manager with json files (thanks to `vim.json`, etc.), and so forth. Very fast and since everything is local, almost everything just works.

On top of that you could probably integrate `nvim-dap` and neovim terminal to integrate remote execution.

Maybe I'd work on that once lua autocommands land. It just feels like the most universal solution, as opposed to rather idiosyncratic solutions that impose other constraints (`distant` seems super cool, but extra binaries and stuff).

In any case, I'm afraid closing. From a `telescope-file-browser` perspective it's just out of scope.
